src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java

```diff
@@ -18,870 +18,924 @@
+                        }
+                        requests.remove(request);
+                        processedRequests++;
+                        break;
+                    }
+                }
+            }
+
+            if (!taskCreated) {
+                declineShort(offer);
+                continue;
+            }
+        }
+        if (processedRequests > 0) {
+            if (!requests.isEmpty()) {
+                LOGGER.info("Created " + processedRequests + " tasks from " + offers.size() + " offers (" + requests.size() + " pending requests).");
+            } else {
+                LOGGER.info("Created " + processedRequests + " tasks from " + offers.size() + " offers.");
+            }
+        } else {
+            if (!requests.isEmpty()) {
+                LOGGER.info("Did not match any of the " + offers.size() + " offers (" + requests.size() + " pending requests)");
+            }
+        }
+        for (Request request: requests) {
+            unmatchedLabels.add(request.request.slaveInfo.getLabelString());
+        }
     }
-  }
 
-  private int getOfferIndex(List<Offer> offers, String recentSlaveId) {
-    int offerIndex = -1;
-    for (Offer offer : offers) {
-      if(offer.getSlaveId().getValue().equals(recentSlaveId)) {
-        offerIndex = offers.indexOf(offer);
-        return offerIndex;
-      }
+    @Override
+    public synchronized void resourceOffers(SchedulerDriver driver, List<Offer> offers) {
+        if (multiThreaded && !isProcessing()) {
+            startProcessing();
+        }
+
+        for (Protos.Offer offer : offers) {
+            boolean queued = offerQueue.offer(offer);
+            if (!queued) {
+                LOGGER.warning("Offer queue is full.");
+                declineShort(offer);
+            } else {
+                LOGGER.info("Queued offer " + offer.getId().getValue());
+            }
+        }
+
+        if (!multiThreaded) {
+            processOffers();
+        }
     }
-    return offerIndex;
-  }
-
-  private boolean matches(Offer offer, Request request) {
-    double cpus = -1;
-    double mem = -1;
-    List<Range> ports = null;
-
-    for (Resource resource : offer.getResourcesList()) {
-      String resourceRole = resource.getRole();
-      String expectedRole = mesosCloud.getRole();
-      if (! (resourceRole.equals(expectedRole) || resourceRole.equals("*"))) {
-        LOGGER.warning("Resource role " + resourceRole +
-            " doesn't match expected role " + expectedRole);
-        continue;
-      }
-      if (resource.getName().equals("cpus")) {
-        if (resource.getType().equals(Value.Type.SCALAR)) {
-          cpus = resource.getScalar().getValue();
+
+    @VisibleForTesting
+    String getFrameworkId() {
+        if (frameworkId != null) {
+            return frameworkId.getValue();
         } else {
-          LOGGER.severe("Cpus resource was not a scalar: " + resource.getType().toString());
+            return NULL_FRAMEWORK_ID;
```

Comment:
Wondering if this line would ever be executed. Is it possible resourceOffers() gets called before registered() ?